tools/gcsbackup/GCSbackup.sh

```diff
@@ -0,0 +1,124 @@
+#!/bin/bash -e
+# Copyright 2016 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+# Please note: GCSbackup should only run inside a GCloud VM instance.
+#
+# GCSbackup is a tool to create and maintain a tagged .tar backup archive
+# for a given local path and copy it to a GCS bucket. It can be configured to
+# maintain a maximum of n backups for specified tag. It automatically
+# deletes older backups with the same tag.
+#
+# On GCS, the .tar is copied to a qualified path that is unique to the VM
+# where this script is running, the username, path, tag, and timestamp.
+
+USAGE='USAGE:
+
+  ./GCSbackup.sh [OPTION]...
+
+OPTIONS:
+
+  -n, --num-backups   Number of backups to keep. Default is 10
+  -b, --bucket        Name of GCS bucket to store backups in. Default is "{project-id}_datalab-backups"
+                      Follow the bucket naming guidelines here: https://cloud.google.com/storage/docs/naming
+  -p, --path          Path to backup. Default is current directory
+  -t, --tag           Tag to make grouping similar backups easy. Default is "backup"
+'
+
+while [[ $# -gt 1 ]]; do
+  key="$1"
+  case $key in
+      -n|--num-backups)
+        num_backups="$2"
+        shift
+        ;;
+      -b|--bucket)
+        gcs_bucket="$2"
+        shift
+        ;;
+      -p|--path)
+        backup_path="$2"
+        shift
+        ;;
+      -t|--tag)
+        tag="$2"
+        shift
+        ;;
+      --default)
+        DEFAULT=YES
+        shift
+        ;;
+      *)
+        echo "Bad arguments found: ${key}"
+        echo "${USAGE}"
+        exit 1
+      ;;
+  esac
+  shift   # skip option value
+done
+
+timestamp=$(date "+%Y%m%d%H%M%S")
+machine_id=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/id" -H "Metadata-Flavor: Google" || echo "")
+project_id=$(gcloud info |tr -d '[]' | awk '/project:/ {print $2}')
+default_bucket="${project_id}_datalab-backups"
```

Comment:
This makes me nervous since bucket names are global (so it may already exist and belong to someone else).

Can we use "${project_id}.appspot.com" for the bucket instead?